src/connection.rs

```diff
@@ -84,145 +84,233 @@ pub trait ClientDisconnect {
 }
 
 /// Astarte device implementation.
-pub struct DeviceConnection<S, C> {
-    interfaces: Arc<RwLock<Interfaces>>,
-    tx: flume::Sender<Result<DeviceEvent, Error>>,
-    pub(crate) client: mpsc::Receiver<ClientMessage>,
-    store: StoreWrapper<S>,
-    connection: C,
+pub struct DeviceConnection<S, C>
+where
+    C: Connection,
+{
+    pub(crate) sender: DeviceSender<S, C::Sender>,
+    pub(crate) receiver: DeviceReceiver<S, C>,
 }
 
-impl<S, C> DeviceConnection<S, C> {
+impl<S, C> DeviceConnection<S, C>
+where
+    C: Connection,
+{
     pub(crate) fn new(
         interfaces: Arc<RwLock<Interfaces>>,
         tx: flume::Sender<Result<DeviceEvent, Error>>,
         client: mpsc::Receiver<ClientMessage>,
         store: StoreWrapper<S>,
         connection: C,
-    ) -> Self {
+        sender: C::Sender,
+    ) -> Self
+    where
+        S: Clone,
+    {
         Self {
-            interfaces,
-            tx,
-            client,
-            store,
-            connection,
+            sender: DeviceSender {
+                interfaces: Arc::clone(&interfaces),
+                client,
+                store: store.clone(),
+                sender,
+            },
+            receiver: DeviceReceiver {
+                interfaces,
+                tx,
+                store,
+                connection,
+            },
         }
     }
+}
 
-    async fn handle_event(
-        &self,
-        interface: &str,
-        path: &str,
-        payload: C::Payload,
-    ) -> Result<Value, crate::Error>
-    where
-        S: PropertyStore,
-        C: Receive + Sync,
-    {
-        let path = MappingPath::try_from(path)?;
+#[async_trait]
+impl<S, C> EventLoop for DeviceConnection<S, C>
+where
+    C: Connection + Reconnect + Receive + Send + Sync + 'static,
+    C::Sender: Register + Publish + Send + Sync + 'static,
+    S: PropertyStore,
+{
+    async fn handle_events(mut self) -> Result<(), crate::Error> {
+        let Self {
+            mut sender,
+            mut receiver,
+        } = self;
 
-        let interfaces = self.interfaces.read().await;
-        let interface = interfaces.get(interface).ok_or_else(|| {
-            warn!("publish on missing interface {interface} ({path})");
-            Error::InterfaceNotFound {
-                name: interface.to_string(),
+        let mut tasks: JoinSet<Result<(), Error>> = JoinSet::new();
+
+        tasks.spawn(async move {
```

Comment:
What do you think to introduce a name for each task? So in case of debugging an error we can distinguish which of task generate the error, or is It already possible with the log that we have?

```suggestion
        task::Builder::new()
            .name(client-receiver-task")
            .spawn(async move {
```

Reply:
We can use a function instead of the async closure. So it will appear in the log traces